From d5ddb9b529e91da310a256049f1522caea45d4c7 Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@nxp.com>
Date: Fri, 2 Nov 2018 17:47:51 +0800
Subject: [PATCH 5010/5242] MLK-20183-2: ASoC: imx-cdnhdmi: reconfigure the
 mclk for HDMI on imx8mq

commit  e8401a450cbff2f7c2251282e36ce0d32a054413 from
https://source.codeaurora.org/external/imx/linux-imx.git

In order to support 44kHz and 48kHz sample rate together, we need to
reconfigure the parent clock of mclk.

Signed-off-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../bindings/sound/imx-audio-cdnhdmi.txt           |    2 +-
 sound/soc/fsl/imx-cdnhdmi.c                        |   13 ++++++++++++-
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/Documentation/devicetree/bindings/sound/imx-audio-cdnhdmi.txt b/Documentation/devicetree/bindings/sound/imx-audio-cdnhdmi.txt
index 2b46b38..6d41217 100644
--- a/Documentation/devicetree/bindings/sound/imx-audio-cdnhdmi.txt
+++ b/Documentation/devicetree/bindings/sound/imx-audio-cdnhdmi.txt
@@ -1,7 +1,7 @@
 Freescale i.MX audio complex with Cadence HDMI
 
 Required properties:
-- compatible : "fsl,imx-audio-cdnhdmi"
+- compatible : "fsl,imx-audio-cdnhdmi", "fsl,imx8mq-evk-cdnhdmi"
 - model : The user-visible name of this sound complex
 - audio-cpu : The phandle of the i.MX SAI controller
 - protocol : 0 is hdmi,  1 is dp.
diff --git a/sound/soc/fsl/imx-cdnhdmi.c b/sound/soc/fsl/imx-cdnhdmi.c
index ba0b32f..8dc9953 100644
--- a/sound/soc/fsl/imx-cdnhdmi.c
+++ b/sound/soc/fsl/imx-cdnhdmi.c
@@ -23,6 +23,7 @@
 #include <sound/soc-dapm.h>
 #include <sound/hdmi-codec.h>
 #include "../../../drivers/gpu/drm/imx/hdp/imx-hdp.h"
+#include "fsl_sai.h"
 
 #define SUPPORT_RATE_NUM 10
 #define SUPPORT_CHANNEL_NUM 10
@@ -98,7 +99,16 @@ static int imx_cdnhdmi_hw_params(struct snd_pcm_substream *substream,
 		return ret;
 	}
 
-	ret = snd_soc_dai_set_sysclk(cpu_dai, 0, 0, tx ? SND_SOC_CLOCK_OUT : SND_SOC_CLOCK_IN);
+
+	if (of_device_is_compatible(dev->of_node,
+				    "fsl,imx8mq-evk-cdnhdmi"))
+		ret = snd_soc_dai_set_sysclk(cpu_dai, FSL_SAI_CLK_MAST1,
+				256 * params_rate(params),
+				SND_SOC_CLOCK_OUT);
+	else
+		ret = snd_soc_dai_set_sysclk(cpu_dai, 0,
+				0,
+				tx ? SND_SOC_CLOCK_OUT : SND_SOC_CLOCK_IN);
 	if (ret) {
 		dev_err(dev, "failed to set cpu sysclk: %d\n", ret);
 		return ret;
@@ -516,6 +526,7 @@ static int imx_cdnhdmi_probe(struct platform_device *pdev)
 }
 
 static const struct of_device_id imx_cdnhdmi_dt_ids[] = {
+	{ .compatible = "fsl,imx8mq-evk-cdnhdmi", },
 	{ .compatible = "fsl,imx-audio-cdnhdmi", },
 	{ /* sentinel */ }
 };
-- 
1.7.9.5

